### DISCUSSION TITLE
How do I configure GitHub for it to verify commits made from my self-hosted Renovate app?

### DISCUSSION THREAD

**Discussion by shenke-adsk** [renovatebot/renovate#22751] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v35.105.0

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I understand that I can configure the Renovate on-prem container to make signed commits by setting the `RENOVATE_GIT_PRIVATE_KEY` env variable. I have made it working (the signing part). However, GitHub Enterprise always shows a `unverified` status on these commits. 

I would like to know what configuration changes I need to make on the GitHub Enterprise part to make it successfully verify the commits made by the bot. 

### What you tried so far.

1. Create a GitHub app on my GitHub Enterprise server using my personal account, deploy the Renovate on-prem container. 
2. Transfer the app owner from my personal account to my organization. I am also a member of this organization. Everything works. 
3. Followed the general instructions to generate a GPG key pair, export the private key, set the env variable `RENOVATE_GIT_PRIVATE_KEY` to use this private key. 
4. Renovate app successfully signed its commits using this key, and I can verify the commits from a computer that trust this key no problem. 
5. However, on the GitHub UI, the commit is shown as `unverified`. Furthermore, it prompts `This user has not uploaded the public key`. 
6. Tried to upload the public key to my personal account's SSH and GPG keys page, key uploaded successfully. But commits are still showing as `unverified`. I am pretty sure the commits made by Renovate are using the same email address as the key I added there. 
7. Searched and read tons of articles, I believe someone has made it working on GitLab. So I think there is something I am missing on the GitHub Enterprise piece. 

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
Set `platformCommit=true`

**Reply by shenke-adsk** [reply]
Hi @rarkins, thanks for the hint. Forgive my ignorance, but I am using the Renovate On-Prem Docker image: https://hub.docker.com/r/whitesource/renovate

Could you please tell me where I can set `platformCommit=true`? 

I am just following the sample code at https://github.com/mend/renovate-on-prem/blob/main/examples/docker-compose-ghe.yml
and I don't have a config file. And the image documentation did not mention anything about using a config file. So far I have been adjusting Renovate behaviour using environment variables described in https://docs.renovatebot.com/self-hosted-configuration/.

**Reply by rarkins** [reply]
You can configure it as `RENOVATE_PLATFORM_COMMIT=true` in env, or as a part of a shared preset if you use those

**Reply by shenke-adsk** [reply]
I did that, but now the commits are no longer signed. I checked the logs, and it appears that the private key was successfully imported, and all the signing configuration is still there. But when I check the commits, the commits have lost the signatures. Also the email address is no longer the one that I set by using env variable `GIT_AUTHOR_EMAIL`. The email address has changed back to the GitHub auto-generated email `22578+renovate-test[bot]@users.noreply.myghehost.com`. 

Adding `RENOVATE_PLATFORM_COMMIT=true` is the only change. 

Right now all the variables that concern commit signing I have are:
```
RENOVATE_PLATFORM_COMMIT=true
GIT_AUTHOR_EMAIL=someveriedmail@mycompany.com
RENOVATE_GIT_PRIVATE_KEY=someprivategpgkey
```

**Reply by rarkins** [reply]
If you enable platform commit then you no longer need gitAuthor or gitPrivateKey. The gitAuthor will be decided by GitHub, and the private key is no longer relevant as it's signed by GitHub. You may need to configure gitIgnoredAuthors to the old address though.

BTW I am assuming that it works the same for GHES as it does for github.com.

**Reply by shenke-adsk** [reply]
So are you saying the RENOVATE_GIT_PRIVATE_KEY approach will not work with GHES? I will make the changes to remove `GIT_AUTHOR_EMAIL` and `RENOVATE_GIT_PRIVATE_KEY` and report what I observe, but I am also curious to know if RENOVATE_GIT_PRIVATE_KEY is a no-go? Have you tried that on github.com?

**Reply by rarkins** [reply]
In theory it should work, but in practice the platformCommit is simpler and my recommendation

**Reply by shenke-adsk** [reply]
Update: I removed the other unnecessary parameters and only kept `RENOVATE_PLATFORM_COMMIT=true`. This time the commit generated from Renovate does not contain any signature. Also, Renovate started to complain: 

```
Renovate will not automatically rebase this PR, because it does not recognize the last commit author and assumes somebody else may have edited the PR.

You can manually request rebase by checking the rebase/retry box above.
```
If this is the only flag that is required for it to work on github.com, then I think for GHES there are some server configurations for administrators to make to enable "automated bot commit signing".

**Reply by rarkins** [reply]
Yes, it's unfortunately sounding that way

**Reply by shenke-adsk** [reply]
Can you point me to the code where Renovate does a platform commit? I want to get a clear understanding of what API it uses and take it to our GitHub support.

**Reply by rarkins** [reply]
FYI this feature has been available in github.com since this announcement: https://github.blog/2019-08-15-commit-signing-support-for-bots-and-other-github-apps/

Current docs here: https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-bots

I don't know whether this is supported for GHES. It doesn't seem to have the same category in https://docs.github.com/en/enterprise-server@3.9/authentication/managing-commit-signature-verification/about-commit-signature-verification

Our low level git is pretty complex but for github you can find it within here: https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/github/index.ts

**Reply by shenke-adsk** [reply]
This can be closed. According to Github support, site admin will have to enable "web commit signing" for GHES, which resolved our issue.

**Comment by shenke-adsk** [comment, marked as answer]
To summarize the things required to make this work:

1. Site admins will need to enable "web commit signing" for GHES. 
2. Use `platformCommit=true` in config file or set `RENOVATE_PLATFORM_COMMIT=true` as env variable if using the Renovate On-prem image. Do not set `RENOVATE_GIT_PRIVATE_KEY`, that won't work with GHES.
